### ISSUE TITLE
Comparison DateTime64 to DateTime / Date

### ISSUE THREAD

**Issue by vladimir-golovchenko** [ClickHouse/ClickHouse#13804]
**Use case**

Comparison between DateTime64 and DateTime / Date types is allowed but the behavior is confusing (it looks like used numerical comparison):

* from one side for zero-precision it works as expected (values compared as datetime-based ones): 

```sql
SELECT toTypeName(dt64) AS dt64_typename, dt64 = dt
FROM 
(
    SELECT 
        toDateTime64(toStartOfInterval(now(), toIntervalSecond(1), 'UTC'), 0, 'UTC') AS dt64,
        toStartOfInterval(now(), toIntervalSecond(1), 'UTC') AS dt
)

/*
┌─dt64_typename────────┬─equals(dt64, dt)─┐
│ DateTime64(0, 'UTC') │                1 │
└──────────────────────┴──────────────────┘
*/
```
* from another side for non-zero-precision its behavior is unexpected if not used explicit type conversion:

```sql
SELECT 
    toTypeName(dt64) AS dt64_typename,
    dt64 = dt,
    toDateTime(dt64) = dt,
    dt64 = toDateTime64(dt, 1, 'UTC')
FROM 
(
    SELECT 
        toDateTime64(toStartOfInterval(now(), toIntervalSecond(1), 'UTC'), 1, 'UTC') AS dt64,
        toStartOfInterval(now(), toIntervalSecond(1), 'UTC') AS dt
)
/*
┌─dt64_typename────────┬─equals(dt64, dt)─┬─equals(toDateTime(dt64), dt)─┬─equals(dt64, toDateTime64(dt, 1, 'UTC'))─┐
│ DateTime64(1, 'UTC') │                0 │                            1 │                                        1 │
└──────────────────────┴──────────────────┴──────────────────────────────┴──────────────────────────────────────────┘
*/
```

**Describe the solution you'd like**
I think this comparison either:
- should be not allowed and throw an exception to ask for user uses explicit conversion
- or make valid implicit conversion relying on datetime/date essence of values

**Comment by den-crane** [comment]
related (duplicate?) https://github.com/ClickHouse/ClickHouse/issues/11222

**Comment by Enmk** [comment]
It looks like this issue was recently fixed with #16952 :

```
SELECT
    n,
    toTypeName(dt64) AS dt64_typename,

    '<',
    dt64 < dt,
    toDateTime(dt64) < dt,
    dt64 < toDateTime64(dt, 1, 'UTC'),
    
    '<=',
    dt64 <= dt,
    toDateTime(dt64) <= dt,
    dt64 <= toDateTime64(dt, 1, 'UTC'),

    '=',
    dt64 = dt,
    toDateTime(dt64) = dt,
    dt64 = toDateTime64(dt, 1, 'UTC'),
    
    '>=',
    dt64 >= dt,
    toDateTime(dt64) >= dt,
    dt64 >= toDateTime64(dt, 1, 'UTC'),
    
    '>',
    dt64 > dt,
    toDateTime(dt64) > dt,
    dt64 > toDateTime64(dt, 1, 'UTC'),

    '!=',
    dt64 != dt,
    toDateTime(dt64) != dt,
    dt64 != toDateTime64(dt, 1, 'UTC')
FROM 
(
    SELECT
        number - 1 as n,
        toDateTime64(toStartOfInterval(now(), toIntervalSecond(1), 'UTC'), 1, 'UTC') + n AS dt64,
        toStartOfInterval(now(), toIntervalSecond(1), 'UTC') AS dt
    FROM system.numbers
    LIMIT 3
)
FORMAT TabSeparated

/*
-1      DateTime64(1, \'UTC\')  <       1       1       1       <=      1       1       1       =       0       0       0       >=      0       0       0       >       0       0       0       !=      1       1       1
0       DateTime64(1, \'UTC\')  <       0       0       0       <=      1       1       1       =       1       1       1       >=      1       1       1       >       0       0       0       !=      0       0       0
1       DateTime64(1, \'UTC\')  <       0       0       0       <=      0       0       0       =       0       0       0       >=      1       1       1       >       1       1       1       !=      1       1       1
*/
```
Could you please confirm?

**Comment by alexey-milovidov** [comment]
@Enmk PR was reverted due to flacky test, please make another PR.
